flexmeasures_client.py

```diff
@@ -335,45 +323,131 @@ def trigger_schedule(self, *args, **fnc_kwargs):
+        # 3. SRW < NOW < B2MS < CTM and SRW < NOW < CTM < B2MS
+        #    Here the priority is to reach the CTM and so not soc-maxima.
+        #
+        # Note that the situation where CTM < NOW is not relevant anymore and is covered by scenario 1.
 
         rounded_now = time_round(self.get_now(), resolution)
+        soc_maxima = []
 
-        # This is when the target SoC cannot be reached before the calendar-item_start,
-        # the start of the relaxation window would have to be in the past. We thus simply start asap: now.
         if start_relaxation_window < rounded_now:
-            start_relaxation_window = rounded_now
+            # This is when the target SoC cannot be reached at the calendar-item_start,
+            # Scenario 3.
+            soc_maxima = []
+            self.log("Strategy for soc_maxima: Priority for calendar target (Scenario 3), no soc_maxima.")
+        else:
+            back_to_max_soc = fnc_kwargs["back_to_max_soc"]
+            if isinstance(back_to_max_soc, datetime):
+                # There is a B2MS
+                minimum_discharge_window = math.ceil((current_soc_kwh - c.CAR_MAX_SOC_IN_KWH) / (c.CHARGER_MAX_DIS_CHARGE_POWER / 1000) * 60)
+                end_minimum_discharge_window = time_round((rounded_now - timedelta(minutes=minimum_discharge_window)), resolution)
+                if end_minimum_discharge_window > back_to_max_soc:
+                    # Scenario A.
+                    back_to_max_soc = end_minimum_discharge_window
+                
+                self.log(f"trigger_schedule, back_to_max_soc: '{back_to_max_soc}'.")
+
+                if back_to_max_soc >= start_relaxation_window:
+                    # Scenario 2.
+                    soc_maxima = [
+                        {
+                            "value": current_soc_kwh,
+                            "datetime": dt.isoformat(),
+                        } for dt in [rounded_now + x * resolution for x in range(0, (start_relaxation_window - rounded_now) // resolution) ]
+                    ]
+                    self.log("Strategy for soc_maxima: Maxima current_soc until Start of relaxation window (Scenario 2).")
+                else:
+                    # Scenario 1.
+                    soc_maxima_higher_max_soc = []
+                    number_of_steps = (back_to_max_soc - rounded_now) // resolution
+                    if number_of_steps > 0:
+                        step_kwh = (current_soc_kwh - c.CAR_MAX_SOC_IN_KWH) / number_of_steps
+                        i = 0
+                        while i < number_of_steps:
+                            d = {
+                                "value": current_soc_kwh - (i * step_kwh),
+                                "datetime": (rounded_now + i * resolution).isoformat()
+                            }
+                            self.log(f"dict: {d}")
+                            soc_maxima_higher_max_soc.append(d)
+                            i += 1
```

Comment:
`soc_maxima_higher_max_soc += [{<dict you assigned to d>} for i in range(number_of_steps)]`

Reply:
Better! More elegant!